# ETL output_steam_games.json 

## Importando librerías 

In [39]:
import json
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import pyarrow as pa
import pyarrow.parquet as pq


## Extracción de data y Exploración 

In [4]:
#extraccion del json
row = [] #lista vacia para ir guardando las filas
with open ("data/output_steam_games.json") as file: #utilizo with para que el archivo se abra y cierre
    for line in file.readlines(): #bucle para ir leyendo filas y luego agregarlas a row
        data = json.loads(line)
        row.append(data)

# genero el dataframe 
games = pd.DataFrame(row)
games

publisher                                 genres   
0                   NaN                                    NaN  \
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title   
0                            NaN                       NaN  \
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date   
0                                                     NaN          NaN  \
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags   
0                                                     NaN  \
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  discount_price   
0                                                     NaN             NaN  \
1                                                     NaN             NaN   
2                                                     NaN             NaN   
3                                                     NaN             NaN   
4                                                     NaN             NaN   
...                                                   ...             ...   
120440  http://steamcommunity.com/app/773640/reviews/?...            1.49   
120441  http://steamcommunity.com/app/733530/reviews/?...            4.24   
120442  http://steamcommunity.com/app/610660/reviews/?...            1.39   
120443  http://steamcommunity.com/app/658870/reviews/?...             NaN   
120444 

In [5]:
games.shape

(120445, 19)

In [6]:
games.describe()

discount_price   items_count
count      225.000000  88310.000000
mean        11.930533     58.353629
std         17.492643    122.312095
min          0.490000      0.000000
25%          1.390000      3.000000
50%          4.190000     26.000000
75%         22.660000     73.000000
max        139.990000   7762.000000

In [7]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       24083 non-null  object 
 1   genres          28852 non-null  object 
 2   app_name        32133 non-null  object 
 3   title           30085 non-null  object 
 4   url             32135 non-null  object 
 5   release_date    30068 non-null  object 
 6   tags            31972 non-null  object 
 7   reviews_url     32133 non-null  object 
 8   discount_price  225 non-null    float64
 9   specs           31465 non-null  object 
 10  price           30758 non-null  object 
 11  early_access    32135 non-null  object 
 12  id              32133 non-null  object 
 13  metascore       2677 non-null   object 
 14  developer       28836 non-null  object 
 15  user_id         88310 non-null  object 
 16  steam_id        88310 non-null  object 
 17  items           88310 non-nul

### Existen columnas que no se utilizan por lo tanto se eliminan con el fin de optimizar el tratamiento de los datos

In [8]:
#elimino las columnas que no estan en el diccionario de datos
steam_games = games.drop(columns = ["user_id", "steam_id","items", "items_count"])
steam_games

publisher                                 genres   
0                   NaN                                    NaN  \
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title   
0                            NaN                       NaN  \
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date   
0                                                     NaN          NaN  \
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags   
0                                                     NaN  \
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  discount_price   
0                                                     NaN             NaN  \
1                                                     NaN             NaN   
2                                                     NaN             NaN   
3                                                     NaN             NaN   
4                                                     NaN             NaN   
...                                                   ...             ...   
120440  http://steamcommunity.com/app/773640/reviews/?...            1.49   
120441  http://steamcommunity.com/app/733530/reviews/?...            4.24   
120442  http://steamcommunity.com/app/610660/reviews/?...            1.39   
120443  http://steamcommunity.com/app/658870/reviews/?...             NaN   
120444 

### Busco nulos y duplicados

In [9]:
nulos_por_columna=steam_games.isnull().sum()
nulos_por_columna


publisher          96362
genres             91593
app_name           88312
title              90360
url                88310
release_date       90377
tags               88473
reviews_url        88312
discount_price    120220
specs              88980
price              89687
early_access       88310
id                 88312
metascore         117768
developer          91609
dtype: int64

In [10]:
Total_nulos= steam_games.isnull().sum().sum()
Total_nulos

1406985

In [11]:
#el dataframe contiene el 80% de datos nulos por lo que se eliminan
steam_games = steam_games.dropna(how="all").reset_index(drop=True)
steam_games

publisher                                             genres   
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]  \
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title   
0           Lost Summoner Kitty       Lost Summoner Kitty  \
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
1      http://steamcommunity.com/app/643980/reviews/?...             NaN   
2      http://steamcommunity.com/app/670290/reviews/?...             NaN   
3      http://steamcommunity.com/app/767400/reviews/?...            0.83   
4      http://steamcommunity.com/app/773570/reviews/?...            1.79   
...                                                  ...             ...   
32130  http://steamcommunity.com/app/773640/reviews/?...            1.49   
32131  http://steamcommunity.com/app/733530/reviews/?...            4.24   
32132  http://steamcommunity.com/app/610660/reviews/?...      

In [12]:
#busco y analizo los duplicados
columnas_para_verificar = ['id']
duplicados_columnas = steam_games[steam_games.duplicated(subset=columnas_para_verificar)]
duplicados_columnas



publisher               genres   
14573                                 Bethesda Softworks             [Action]  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name   
14573                Wolfenstein II: The New Colossus  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                title   
14573                Wolfenstein II: The New Colossus  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26  \
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags   
14573  [Action, FPS, Gore, Violent, Alternate History...  \
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  discount_price   
14573  http://steamcommunity.com/app/612880/reviews/?...             NaN  \
30961                                                NaN             NaN   

                                                   specs  price early_access   
14573  [Single-player, Steam Achievements, Full contr...  59.99        False  \
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id metascore                                   developer  
14573  612880        86                               Machine Games  
30961     NaN        91  Rocksteady Studios,Feral Interactive (Mac)

In [13]:
steam_games[steam_games["developer"]== "Machine Games"]

publisher    genres   
12237  Bethesda Softworks  [Action]  \
13318  Bethesda Softworks  [Action]   
13894  Bethesda Softworks  [Action]   
14573  Bethesda Softworks  [Action]   
28996  Bethesda Softworks  [Action]   

                                                app_name   
12237  Wolfenstein II: The Freedom Chronicles - Seaso...  \
13318  Wolfenstein II: The Freedom Chronicles - Episo...   
13894                   Wolfenstein II: The New Colossus   
14573                   Wolfenstein II: The New Colossus   
28996                         Wolfenstein: The New Order   

                                                   title   
12237  Wolfenstein II: The Freedom Chronicles - Seaso...  \
13318  Wolfenstein II: The Freedom Chronicles - Episo...   
13894                   Wolfenstein II: The New Colossus   
14573                   Wolfenstein II: The New Colossus   
28996                         Wolfenstein: The New Order   

                                                     url release_date   
12237  http://store.steampowered.com/app/650410/Wolfe...   2017-11-07  \
13318  http://store.steampowered.com/app/624620/Wolfe...   2017-12-13   
13894          http://store.steampowered.com/app/612880/   2017-10-26   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   
28996  http://store.steampowered.com/app/201810/Wolfe...   2014-05-19   

                                                    tags   
12237                            [Gore, Action, Violent]  \
13318                                  [Action, Violent]   
13894  [Action, FPS, Gore, Violent, Alternate History...   
14573  [Action, FPS, Gore, Violent, Alternate History...   
28996  [FPS, Action, Shooter, Alternate History, Firs...   

                                             reviews_url  discount_price   
12237  http://steamcommunity.com/app/650410/reviews/?...             NaN  \
13318  http://steamcommunity.com/app/624620/reviews/?...             NaN   
13894  http://steamcommunity.com/app/612880/reviews/?...             NaN   
14573  http://steamcommunity.com/app/612880/reviews/?...             NaN   
28996  http://steamcommunity.com/app/201810/reviews/?...             NaN   

                                                   specs  price early_access   
12237  [Single-player, Downloadable Content, Full con...  24.99        False  \
13318  [Single-player, Downloadable Content, Steam Ac...   9.99        False   
13894  [Single-player, Steam Achievements, Full contr...  59.99        False   
14573  [Single-player, Steam Achievements, Full contr...  59.99        False   
28996  [Single-player, Steam Achievements, Partial Co...  19.99        False   

           id metascore      developer  
12237  650410       NaN  Machine Games  
13318  624620       NaN  Machine Games  
13894  612880        86  Machine Games  
14573  612880        86  Machine Games  
28996  201810        81  Machine Games

#### La fila duplicada 14573 es idéntica a la 13894.. por lo que podemos eliminarla

In [14]:
steam_games[steam_games["developer"]== "Rocksteady Studios,Feral Interactive (Mac)"]

publisher               genres   
1068   Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   
31617  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name   
1068   Batman: Arkham City - Game of the Year Edition  \
30961  Batman: Arkham City - Game of the Year Edition   
31617  Batman: Arkham Asylum Game of the Year Edition   

                                                title   
1068   Batman: Arkham City - Game of the Year Edition  \
30961  Batman: Arkham City - Game of the Year Edition   
31617  Batman: Arkham Asylum Game of the Year Edition   

                                                     url release_date   
1068   http://store.steampowered.com/app/200260/Batma...   2012-09-07  \
30961           http://store.steampowered.com/app/200260   2012-09-07   
31617  http://store.steampowered.com/app/35140/Batman...   2010-03-26   

                                                    tags   
1068   [Action, Open World, Batman, Adventure, Stealt...  \
30961  [Action, Open World, Batman, Adventure, Stealt...   
31617  [Action, Batman, Stealth, Adventure, Third Per...   

                                             reviews_url  discount_price   
1068   http://steamcommunity.com/app/200260/reviews/?...             NaN  \
30961                                                NaN             NaN   
31617  http://steamcommunity.com/app/35140/reviews/?b...             NaN   

                                                   specs  price early_access   
1068   [Single-player, Steam Achievements, Steam Trad...  19.99        False  \
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   
31617  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id metascore                                   developer  
1068   200260        91  Rocksteady Studios,Feral Interactive (Mac)  
30961     NaN        91  Rocksteady Studios,Feral Interactive (Mac)  
31617   35140        91  Rocksteady Studios,Feral Interactive (Mac)

#### La fila duplicada 30961, es idéntica a la 1068; pero contiene el dato de id nulo, mientras que la fila 1068 posee el id completo 200260. Por lo que se elimina la fila duplicada 30961 

In [15]:
#elimino columnas con id duplicadas
duplicados_col = [14573, 30961]
steam_games= steam_games.drop(duplicados_col)

In [16]:
columnas_para_verificar = ['id']
duplicados_columnas = steam_games[steam_games.duplicated(subset=columnas_para_verificar)]
duplicados_columnas

Empty DataFrame
Columns: [publisher, genres, app_name, title, url, release_date, tags, reviews_url, discount_price, specs, price, early_access, id, metascore, developer]
Index: []

In [17]:
steam_games.shape

(32133, 15)

## Transformación de los datos 

In [19]:
# conozco que tipos de datos tiene el dataset para saber cual debo modificar
tipos_de_datos=steam_games.dtypes
tipos_de_datos

publisher          object
genres             object
app_name           object
title              object
url                object
release_date       object
tags               object
reviews_url        object
discount_price    float64
specs              object
price              object
early_access       object
id                 object
metascore          object
developer          object
dtype: object

#### EL dataset aún es un archivo json, por lo que el tipo de dato object , lo que puede ser datos anidados con formato clave - valor. Entonces a través de un bucle se analiza columna por columna y se almacenan en un diccionario vacio para luego transformarlo en un dataset y alli poder ver el tipo de datos de las columnas.

In [20]:
tipo_data = {"columna":[],"tipos_de_datos":[]} #genero un diccionario vacio para ir almacenando lo que genere el bucle

for columna in steam_games.columns: #un bucle que va recorriendo 
    tipo_data["columna"].append(columna)
    tipo_data["tipos_de_datos"].append(steam_games[columna].apply(type).unique())

analisis= pd.DataFrame(tipo_data)
analisis

columna                                   tipos_de_datos
0        publisher                 [<class 'str'>, <class 'float'>]
1           genres                [<class 'list'>, <class 'float'>]
2         app_name                 [<class 'str'>, <class 'float'>]
3            title                 [<class 'str'>, <class 'float'>]
4              url                                  [<class 'str'>]
5     release_date                 [<class 'str'>, <class 'float'>]
6             tags                [<class 'list'>, <class 'float'>]
7      reviews_url                 [<class 'str'>, <class 'float'>]
8   discount_price                                [<class 'float'>]
9            specs                [<class 'list'>, <class 'float'>]
10           price                 [<class 'float'>, <class 'str'>]
11    early_access                                 [<class 'bool'>]
12              id                 [<class 'str'>, <class 'float'>]
13       metascore  [<class 'float'>, <class 'int'>, <class 'str'>]
14       developer                 [<class 'str'>, <class 'float'>]

### Transformación de la columna "release_date"

#### Es necesario que pueda extraerse el año del dato fecha ; por lo que se transforma la columna "release_date", generando una con el año de estreno

In [21]:
#composición de la columna
steam_games["release_date"]

0        2018-01-04
1        2018-01-04
2        2017-07-24
3        2017-12-07
4               NaN
            ...    
32130    2018-01-04
32131    2018-01-04
32132    2018-01-04
32133    2017-09-02
32134           NaN
Name: release_date, Length: 32133, dtype: object

In [22]:
steam_games["release_date"].value_counts()

release_date
2012-10-16    100
2017-08-31     92
2017-09-26     89
2017-06-21     82
2017-07-25     78
             ... 
1988-04-16      1
2013-08-24      1
2011-05-07      1
2010-08-21      1
2018-10-01      1
Name: count, Length: 3582, dtype: int64

In [23]:
# Extraer el año de la columna "release_date" y reemplazar los valores nulos con "sin dato"
steam_games['Año_estreno'] = steam_games['release_date'].str.extract(r'(\d{4})').fillna('sin dato')

steam_games

publisher                                             genres   
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]  \
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title   
0           Lost Summoner Kitty       Lost Summoner Kitty  \
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
1      http://steamcommunity.com/app/643980/reviews/?...             NaN   
2      http://steamcommunity.com/app/670290/reviews/?...             NaN   
3      http://steamcommunity.com/app/767400/reviews/?...            0.83   
4      http://steamcommunity.com/app/773570/reviews/?...            1.79   
...                                                  ...             ...   
32130  http://steamcommunity.com/app/773640/reviews/?...            1.49   
32131  http://steamcommunity.com/app/733530/reviews/?...            4.24   
32132  http://steamcommunity.com/app/610660/reviews/?...      

### Transformación de la columna "price"

#### La columna "price", posee valores nulos que deben ser reemplazados por valores numéricos de tipo flotante. Toda la columna debe ser dato de tipo flotante.

In [24]:
#observo la composición de la columna
steam_games["price"]

0                4.99
1        Free To Play
2        Free to Play
3                0.99
4                2.99
             ...     
32130            1.99
32131            4.99
32132            1.99
32133            4.99
32134            4.99
Name: price, Length: 32133, dtype: object

In [25]:
steam_games["price"].value_counts()

price
4.99           4278
9.99           3902
2.99           3429
0.99           2607
1.99           2541
               ... 
74.76             1
32.99             1
14.95             1
26.99             1
Free to Use       1
Name: count, Length: 162, dtype: int64

In [26]:
# se define una función para poder reemplazar los datos string de la columna "price" por flotantes 0.0

def cambio_a_float(valor):

    if pd.isna(valor): #esta parte de la función es para el caso de que no exista ningun string o float en el campo
        return 0.0
    try:
        flotante = float(valor) #si el valor es float lo conservo
        return flotante
    except (ValueError, TypeError): #si el valor es diferente retorno 0.0
        return 0.0
    
#Aplico esta función a la columna precio
steam_games["price"] = steam_games["price"].apply(cambio_a_float)


In [27]:
#corroboro el cambio de tipo de dato
steam_games["price"]

0        4.99
1        0.00
2        0.00
3        0.99
4        2.99
         ... 
32130    1.99
32131    4.99
32132    1.99
32133    4.99
32134    4.99
Name: price, Length: 32133, dtype: float64

In [28]:
#buscamos outliers en precio... aunque se puede comprender que los valores resultarán muy diferentes, debido a ofertas o juegos gratis.

# Se calcula el IQR (Rango Intercuartil)
Q1 = steam_games['price'].quantile(0.25)
Q3 = steam_games['price'].quantile(0.75)
IQR = Q3 - Q1
print(Q1)
print(Q3)
print(IQR)

1.99
9.99
8.0


In [29]:
#Se defiene el umbral de outliers
umbral_superior = Q3 + 1.5 * IQR
umbral_inferior = Q1 - 1.5 * IQR
print(umbral_superior)
print(umbral_inferior)

21.990000000000002
-10.01


In [30]:
#Se identifican los outliers
outliers_superiores = steam_games[steam_games['price'] > umbral_superior]
outliers_inferiores = steam_games[steam_games['price'] < umbral_inferior]
print(outliers_superiores)
print(outliers_inferiores)

                              publisher   
8                Poppermost Productions  \
18               Poppermost Productions   
51                                  NaN   
59                               SimBin   
234                              NCsoft   
...                                 ...   
31938  Firedog Creative Company Limited   
31956                               NaN   
31959                 CarX Technologies   
32070                  DreamInCode B.V.   
32098                        Koch Media   

                                                  genres   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                                   NaN   
59                                              [Racing]   
234                         [Massively Multiplayer, RPG]   
...                                                  ...   
31938                                  [RPG, Simulation]   
319

In [31]:
outliers_inferiores

Empty DataFrame
Columns: [publisher, genres, app_name, title, url, release_date, tags, reviews_url, discount_price, specs, price, early_access, id, metascore, developer, Año_estreno]
Index: []

In [32]:

outliers_superiores

publisher   
8                Poppermost Productions  \
18               Poppermost Productions   
51                                  NaN   
59                               SimBin   
234                              NCsoft   
...                                 ...   
31938  Firedog Creative Company Limited   
31956                               NaN   
31959                 CarX Technologies   
32070                  DreamInCode B.V.   
32098                        Koch Media   

                                                  genres   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                                   NaN   
59                                              [Racing]   
234                         [Massively Multiplayer, RPG]   
...                                                  ...   
31938                                  [RPG, Simulation]   
31956                                                NaN   
31959  [Massively Multiplayer, Racing, Simulation, Sp...   
32070  [Action, Adventure, Casual, Indie, Simulation,...   
32098                      [Utilities, Video Production]   

                                             app_name   
8                       SNOW - All Access Legend Pass  \
18                               SNOW - Lifetime Pack   
51                   Medicalholodeck Personal Version   
59                               RACE - The WTCC Game   
234                                Guild Wars Trilogy   
...                                               ...   
31938                                           愛神餐館2   
31956  Three Kingdoms VR - Jade Knight (光之三國VR - 青龍騎)   
31959               CarX Drift Racing Online - Deluxe   
32070                                       Mesozoica   
32098                       CyberLink YouCam 7 Deluxe   

                                   title   
8          SNOW - All Access Legend Pass  \
18                  SNOW - Lifetime Pack   
51                                   NaN   
59                  RACE - The WTCC Game   
234                   Guild Wars Trilogy   
...                                  ...   
31938                              愛神餐館2   
31956                                NaN   
31959  CarX Drift Racing Online - Deluxe   
32070                          Mesozoica   
32098          CyberLink YouCam 7 Deluxe   

                                                     url release_date   
8      http://store.steampowered.com/app/774278/SNOW_...   2018-01-04  \
18     http://store.steampowered.com/app/774279/SNOW_...   2018-01-03   
51     http://store.steampowered.com/app/769350/Medic...          NaN   
59     http://store.steampowered.com/app/4230/RACE__T...   2006-11-24   
234    http://store.steampowered.com/app/29570/Guild_...   2008-09-25   
...                                                  ...          ...   
31938        http://store.steampowered.com/app/769680/2/   2017-12-29   
31956  http://store.steampowered.com/app/731120/Three...          NaN   
31959  http://store.steampowered.com/app/776950/CarX_...   2017-12-28   
32070  http://store.steampowered.com/app/719590/Mesoz...   2018-01-02   
32098  http://store.steampowered.com/app/755310/Cyber...   2017-12-26   

                                                    tags   
8              [Free to Play, Indie, Simulation, Sports]  \
18             [Free to Play, Indie, Simulation, Sports]   
51                                           [Education]   
59                                  [Racing, Simulation]   
234            [RPG, Massively Multiplayer, MMORPG, PvP]   
...                                                  ...   
31938                           [RPG, Simulation, Anime]   
31956               [Action, Casual, Simulation, Sports]   
31959  [Massively Multiplayer, Simulation, Racing, Sp...   
32070  [Early Access, Simulation, Strategy, Casual, I...   
32098                      [Utilities, Vi

### Transformación de la columna "genres"

#### Terminada la transformación de los datos numéricos, se trata la columna "genres"

In [33]:
steam_games["genres"][0]

['Action', 'Casual', 'Indie', 'Simulation', 'Strategy']

In [34]:
#con explode hago que la columna genres tome cada valor que tiene almacenado y lo devuelva en un campo nuevo pero conservando el mismo id
steam_games = steam_games.explode("genres")
steam_games = steam_games.dropna(subset=["genres"])
steam_games

publisher      genres             app_name                title   
0         Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty  \
0         Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   
...             ...         ...                  ...                  ...   
32132  Laush Studio       Indie        Russian Roads        Russian Roads   
32132  Laush Studio      Racing        Russian Roads        Russian Roads   
32132  Laush Studio  Simulation        Russian Roads        Russian Roads   
32133      SIXNAILS      Casual  EXIT 2 - Directions  EXIT 2 - Directions   
32133      SIXNAILS       Indie  EXIT 2 - Directions  EXIT 2 - Directions   

                                                     url release_date   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
...                                                  ...          ...   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   

                                                    tags   
0          [Strategy, Action, Indie, Casual, Simulation]  \
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
...                                                  ...   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
...                                                  ...             ...   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   

                                                   specs  price early_access   
0                                        [Single-player]   4.99        False  \
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
...                                                  ...    ...          ...   
32132  [Single

#### Procedo a completar los datos que se encuentran nulos o vacios, para el caso de los datos de tipo string del dataset 

In [35]:
#busco espacios vacios o sin datos 
steam_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71551 entries, 0 to 32133
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       59112 non-null  object 
 1   genres          71551 non-null  object 
 2   app_name        71549 non-null  object 
 3   title           71549 non-null  object 
 4   url             71551 non-null  object 
 5   release_date    71498 non-null  object 
 6   tags            71505 non-null  object 
 7   reviews_url     71551 non-null  object 
 8   discount_price  580 non-null    float64
 9   specs           70994 non-null  object 
 10  price           71551 non-null  float64
 11  early_access    71551 non-null  object 
 12  id              71551 non-null  object 
 13  metascore       5467 non-null   object 
 14  developer       71201 non-null  object 
 15  Año_estreno     71551 non-null  object 
dtypes: float64(2), object(14)
memory usage: 9.3+ MB


In [36]:
#inicio un archivo con las columnas que tengo que reemplazar nulos
archivo_columnas=["publisher", "app_name","title","developer"]

#genero un dataframe con las columnas y sus nulos reemplazados por "Sin Datos"
nulos= steam_games[archivo_columnas].fillna("Sin Datos")

#elimino las colummas a reemplazar y concateno ambos dataframes 
steam_games = pd.concat([steam_games.drop(archivo_columnas, axis=1), nulos], axis=1)
steam_games

genres                                                url   
0          Action  http://store.steampowered.com/app/761140/Lost_...  \
0          Casual  http://store.steampowered.com/app/761140/Lost_...   
0           Indie  http://store.steampowered.com/app/761140/Lost_...   
0      Simulation  http://store.steampowered.com/app/761140/Lost_...   
0        Strategy  http://store.steampowered.com/app/761140/Lost_...   
...           ...                                                ...   
32132       Indie  http://store.steampowered.com/app/610660/Russi...   
32132      Racing  http://store.steampowered.com/app/610660/Russi...   
32132  Simulation  http://store.steampowered.com/app/610660/Russi...   
32133      Casual  http://store.steampowered.com/app/658870/EXIT_...   
32133       Indie  http://store.steampowered.com/app/658870/EXIT_...   

      release_date                                               tags   
0       2018-01-04      [Strategy, Action, Indie, Casual, Simulation]  \
0       2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
0       2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
0       2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
0       2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
...            ...                                                ...   
32132   2018-01-04                        [Indie, Simulation, Racing]   
32132   2018-01-04                        [Indie, Simulation, Racing]   
32132   2018-01-04                        [Indie, Simulation, Racing]   
32133   2017-09-02  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32133   2017-09-02  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   

                                             reviews_url  discount_price   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49  \
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
0      http://steamcommunity.com/app/761140/reviews/?...            4.49   
...                                                  ...             ...   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32132  http://steamcommunity.com/app/610660/reviews/?...            1.39   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   
32133  http://steamcommunity.com/app/658870/reviews/?...             NaN   

                                                   specs  price early_access   
0                                        [Single-player]   4.99        False  \
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
0                                        [Single-player]   4.99        False   
...                                                  ...    ...          ...   
32132  [Single-player, Steam Achievements, Steam Trad...   1.99        False   
32132  [Single-player, Steam Achievements, Steam Trad...   1.99        False   
32132  [Single-player, Steam Achievements, Steam Trad...   1.99        False   
32133   [Single-player, Steam Achievements, Steam Cloud]   4.99        False   
32133   [Single-player, Steam Achievements, Steam Cloud]   4.99        False   

           id metascore Año_estreno     publisher             app_name   
0      761140       NaN        2018     Kotoshiro  Lost Summoner Kitty  \
0      761140       NaN        2018     Kotoshiro  Lost Summoner Kitty   
0      761140       NaN        2018     Kotoshiro  Lost Summoner Kitty   
0      761140       NaN        2018     Kotoshiro  Lost Summoner Kitty   
0      761140       NaN  

In [37]:
steam_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71551 entries, 0 to 32133
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          71551 non-null  object 
 1   url             71551 non-null  object 
 2   release_date    71498 non-null  object 
 3   tags            71505 non-null  object 
 4   reviews_url     71551 non-null  object 
 5   discount_price  580 non-null    float64
 6   specs           70994 non-null  object 
 7   price           71551 non-null  float64
 8   early_access    71551 non-null  object 
 9   id              71551 non-null  object 
 10  metascore       5467 non-null   object 
 11  Año_estreno     71551 non-null  object 
 12  publisher       71551 non-null  object 
 13  app_name        71551 non-null  object 
 14  title           71551 non-null  object 
 15  developer       71551 non-null  object 
dtypes: float64(2), object(14)
memory usage: 9.3+ MB


## Guardo el dataset como csv

In [68]:
guardar = "data/steam_games.csv"

steam_games.to_csv(guardar, index=False, encoding="utf-8")

## Transformo csv a parquet para una optimización de los archivos

In [40]:
#Transformo el archivo csv a parquet
#Leo el archivo csv
steam_games= pd.read_csv("data/steam_games.csv") 

#Indico donde quiero guardar el parquet y con que nombre
output_file= "data/steam_games.parquet"

#Transformo a traves de una tabla el archivo csv en parquet
table = pa.Table.from_pandas(steam_games)
pq.write_table(table,output_file)